Sources/PackageLoading/PackageDescription4Loader.swift

```diff
@@ -342,7 +343,22 @@ extension PackageDependencyDescription {
         let requirement = try Requirement(v4: json.get("requirement"))
         let location = try fixLocation(json.get("url"), requirement: requirement)
         let name: String? = json.get("name")
-        self.init(name: name, url: location, requirement: requirement)
+        
+        // Provides better feedback when mistakingly using url: for a dependancy that
+        // is a local package. Still allows for using url with a local package that has
+        // also been initialized by git
+        if let validPath = try? AbsolutePath(validating: url), localFileSystem.exists(validPath) {
+            do {
+                let gitRepoProvider = GitRepositoryProvider()
+                try gitRepoProvider.callGit("-C", url, "rev-parse", "--git-dir", repository: RepositorySpecifier(url: url))
+            } catch let error as GitCloneError {
+                throw GitCloneError(repository: RepositorySpecifier(url: url),
```

Comment:
throw StringError, GitCloneError is reserved to the Git modules

Reply:
I don't think we should be throwing these errors during package loading. In principle, developers can add any number of "broken" dependencies to their package that are working on their local machine as long as they are not used by any of the package's products. I am not sure we should be breaking that contract simply to get a better error message. A solution to this could be capturing the error here and keeping it around in the model and we would emit them only at the time the dependency is actually being used.

The other concern I have is the direct use of `GitRepositoryProvider` which will just be an obstacle if we ever want to support other kinds of repositories. I think this should instead still be in the repository provider and we will need a public API that let's us call into this functionality. That would also alleviate Tom's concern about the `GitCloneError`.

Reply:
great points 👍